libgoal/lockedFileLinux.go

```diff
@@ -19,53 +19,68 @@
 package libgoal
 
 import (
+	"golang.org/x/sys/unix"
 	"io"
 	"os"
-	"syscall"
-
-	"golang.org/x/sys/unix"
 )
 
 type linuxLocker struct {
+	setLockWait int
 }
 
 // makeLocker create a unix file locker.
-// note that the desired way is to use the OFD locker, which locks on the file descriptor level.
-// falling back to the non-OFD lock would allow obtaining two locks by the same process. If this becomes
-// and issue, we might want to use flock, which wouldn't work across NFS.
+// Note that the desired way is to use the OFD locker, which locks on the file descriptor level.
+// Since older kernels (Linux kernel < 3.15) do not support OFD, we fall back to non-OFD in that case.
+// Falling back to the non-OFD lock would allow obtaining two locks by the same process. If this becomes
+// and issue, we might want to use flock, which wouldn't work across NFS on older Linux kernels.
 func makeLocker() *linuxLocker {
 	locker := &linuxLocker{}
+
+	// Check whether F_OFD_SETLKW is supported
+	getlk := unix.Flock_t{Type: unix.F_RDLCK}
+	err := unix.FcntlFlock(0, unix.F_OFD_GETLK, &getlk)
+	if err == nil {
+		locker.setLockWait = unix.F_OFD_SETLKW
+	} else if err == unix.EINVAL {
+		// The command F_OFD_SETLKW is not available
+		// Fall back to non-OFD locks
+		locker.setLockWait = unix.F_SETLKW
+	} else {
+		// Another unknown error occurred, panic
+		panic(err)
```

Comment:
Panicing is something that we're trying to eliminate. It's not where we want to go.
Instead, return a nil and add logic to the caller so that it could make the right decision of what to do.
i.e. make this function return a (*linuxLocker, error) and return (nil, errUnableToCreateFileLock )

Reply:
fixed